### ISSUE TITLE
TF2 keras.models.load_model fails with custom metrics (both h5 and tf format)

### ISSUE THREAD

**Issue by durandg12** [tensorflow/tensorflow#34068]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.13.6
- TensorFlow installed from (source or binary): pip install tensorflow==2.0.0-beta1
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: v3.6.7:6ec5cf24b7, Oct 20 2018, 03:02:14

**Describe the current behavior**

I have a custom metric in my model and using `tf.keras.models.load_model` with `compile=True` after saving it results in an error in almost all cases, whereas I use the `custom_objects` argument according to [the documentation](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/models/load_model).

I tried to pass my custom metric with two strategies: by passing a custom function `custom_accuracy` to the `tf.keras.Model.compile` method, or by subclassing the `MeanMetricWrapper` class and giving an instance of my subclass named `CustomAccuracy` to `tf.keras.Model.compile`.

I also tried the two different saving format available: h5 and tf. Here are my results:

- with tf format:
    - with custom function: 
    fail with `ValueError` message `Unknown metric function:custom_accuracy`
    - with subclassed metric:
    fail with `ValueError` message `Unknown metric function: CustomAccuracy`
- with h5 format:
    - with custom function: 
    success
    - with subclassed metric:
    fail with `TypeError` message `must be str, not ABCMeta`

Note that given the complete error logs (see below), the error with h5 format and subclassed metric is in fact the same as the error with the tf format. The `TypeError` occurs when the code tries to raise the `ValueError`.

**Describe the expected behavior**

This should not fail in any case, except if I am using the `custom_objects` argument wrong. The documentation could be a little expanded on that matter by the way.

**Code to reproduce the issue**
```
import tensorflow as tf

print('Using Tensorflow version {} (git version {})'.format(tf.version.VERSION, tf.version.GIT_VERSION))

from tensorflow.python.keras.metrics import MeanMetricWrapper
from tensorflow.python.keras.metrics import accuracy

def custom_accuracy(y_true, y_pred):
    return accuracy(y_true, y_pred)

class CustomAccuracy(MeanMetricWrapper):

    def __init__(self, **kwargs):
        super(CustomAccuracy, self).__init__(custom_accuracy, **kwargs)
        
def make_model():
    inp = tf.keras.Input(shape=(2,))
    x = tf.keras.layers.Dense(4)(inp)
    return tf.keras.Model(inp, x)


for save_format in ['tf', 'h5']:
    
    print("\nTrying with save_format='{}':\n".format(save_format))
    
    model_with_function = make_model()
    model_with_function.compile(loss='mse', metrics=[custom_accuracy])
    model_with_function.save('/tmp/model_with_function' + '.' + save_format, 
                             save_format=save_format)
    
    try:
        new_model = tf.keras.models.load_model('/tmp/model_with_function' + '.' + save_format, 
                                               custom_objects={'custom_accuracy': custom_accuracy}, 
                                               compile=True)
        print("model_with_function loaded with the following metrics:")
        print(new_model.metrics)
    except Exception as e:
        print("model_with_function not loaded with the following error:")
        print(type(e))
        print(e)
        
    model_with_subclass = make_model()
    model_with_subclass.compile(loss='mse', metrics=[CustomAccuracy()])
    model_with_subclass.save('/tmp/model_with_subclass' + '.' + save_format, 
                             save_format=save_format)
    
    try:
        new_model = tf.keras.models.load_model('/tmp/model_with_subclass' + '.' + save_format, 
                                               custom_objects={'CustomAccuracy': CustomAccuracy}, 
                                               compile=True)
        print("model_with_subclass loaded with the following metrics:")
        print(new_model.metrics)
    except Exception as e:
        print("model_with_subclass not loaded with the following error:")
        print(type(e))
        print(e)
```

**Other info / logs**
The logs are the same in the 3 error cases (to get them with the code above, just add `raise`at the end of the `except` blocks):
```
<ipython-input-14-ac0a72b492dc> in <module>
     48         new_model = tf.keras.models.load_model('/tmp/model_with_subclass' + '.' + save_format, 
     49                                                custom_objects={'CustomAccuracy': CustomAccuracy},
---> 50                                                compile=True)
     51         print("model_with_function loaded with the following metrics:")
     52         print(new_model.metrics)

/path/to/tensorflow_core/python/keras/saving/save.py in load_model(filepath, custom_objects, compile)
    148   if isinstance(filepath, six.string_types):
    149     loader_impl.parse_saved_model(filepath)
--> 150     return saved_model_load.load(filepath, compile)
    151 
    152   raise IOError(

/path/to/tensorflow_core/python/keras/saving/saved_model/load.py in load(path, compile)
     91     if model._training_config is not None:  # pylint: disable=protected-access
     92       model.compile(**saving_utils.compile_args_from_training_config(
---> 93           model._training_config))  # pylint: disable=protected-access
     94 
     95   return model

/path/to/tensorflow_core/python/training/tracking/base.py in _method_wrapper(self, *args, **kwargs)
    455     self._self_setattr_tracking = False  # pylint: disable=protected-access
    456     try:
--> 457       result = method(self, *args, **kwargs)
    458     finally:
    459       self._self_setattr_tracking = previous_value  # pylint: disable=protected-access

/path/to/tensorflow_core/python/keras/engine/training.py in compile(self, optimizer, loss, metrics, loss_weights, sample_weight_mode, weighted_metrics, target_tensors, distribute, **kwargs)
    354     with K.get_graph().as_default():
    355       # Save all metric attributes per output of the model.
--> 356       self._cache_output_metric_attributes(metrics, weighted_metrics)
    357 
    358       # Set metric attributes on model.

/path/to/tensorflow_core/python/keras/engine/training.py in _cache_output_metric_attributes(self, metrics, weighted_metrics)
   1899         output_shapes.append(output.shape.as_list())
   1900     self._per_output_metrics = training_utils.collect_per_output_metric_info(
-> 1901         metrics, self.output_names, output_shapes, self.loss_functions)
   1902     self._per_output_weighted_metrics = (
   1903         training_utils.collect_per_output_metric_info(

/path/to/tensorflow_core/python/keras/engine/training_utils.py in collect_per_output_metric_info(metrics, output_names, output_shapes, loss_fns, is_weighted)
    811     metrics_dict = OrderedDict()
    812     for metric in metrics:
--> 813       metric_name = get_metric_name(metric, is_weighted)
    814       metric_fn = get_metric_function(
    815           metric, output_shape=output_shapes[i], loss_fn=loss_fns[i])

/path/to/tensorflow_core/python/keras/engine/training_utils.py in get_metric_name(metric, weighted)
    985       return metric
    986 
--> 987     metric = metrics_module.get(metric)
    988     return metric.name if hasattr(metric, 'name') else metric.__name__
    989   else:

/path/to/tensorflow_core/python/keras/metrics.py in get(identifier)
   2855 def get(identifier):
   2856   if isinstance(identifier, dict):
-> 2857     return deserialize(identifier)
   2858   elif isinstance(identifier, six.string_types):
   2859     return deserialize(str(identifier))

/path/to/tensorflow_core/python/keras/metrics.py in deserialize(config, custom_objects)
   2849       module_objects=globals(),
   2850       custom_objects=custom_objects,
-> 2851       printable_module_name='metric function')
   2852 
   2853 

/path/to/tensorflow_core/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    178     config = identifier
    179     (cls, cls_config) = class_and_config_for_serialized_keras_object(
--> 180         config, module_objects, custom_objects, printable_module_name)
    181 
    182     if hasattr(cls, 'from_config'):

/path/to/tensorflow_core/python/keras/utils/generic_utils.py in class_and_config_for_serialized_keras_object(config, module_objects, custom_objects, printable_module_name)
    163     cls = module_objects.get(class_name)
    164     if cls is None:
--> 165       raise ValueError('Unknown ' + printable_module_name + ': ' + class_name)
    166   return (cls, config['config'])
    167 
```

**Comment by ravikyram** [comment]
@durandg12 
I tried reproducing the code in colab using TF 2.0 beta1, TF 2.0 and i am seeing different error messages. Kindly , provide minimal stand alone reproducible code,it  helps us in localizing the issue faster.Please, find the gist [here.](https://colab.sandbox.google.com/gist/ravikyram/1780a9fe7e8d0e39cf52ba6e973d22e6/untitled342.ipynb) Thanks!

**Comment by durandg12** [comment]
@ravikyram 
I have looked at your gist.
The error messages in your gist for tf2.0.0 are exactly the same as mine.
There is also a deprecation warning that I have too but that I hadn't copied in my first message because it didn't seem relevant. 

For tf2.0.0-beta1 the error message is effectively different but it comes from the `compile` method because I call it without an `optimizer` argument. I have added `optimizer='adam'` in my `compile` call and now the output is the same for 2.0.0 and 2.0.0-beta1. So the code now looks like this:
```
import tensorflow as tf

print('Using Tensorflow version {} (git version {})'.format(tf.version.VERSION, tf.version.GIT_VERSION))

from tensorflow.python.keras.metrics import MeanMetricWrapper
from tensorflow.python.keras.metrics import accuracy

def custom_accuracy(y_true, y_pred):
    return accuracy(y_true, y_pred)

class CustomAccuracy(MeanMetricWrapper):

    def __init__(self, **kwargs):
        super(CustomAccuracy, self).__init__(custom_accuracy, **kwargs)
        
def make_model():
    inp = tf.keras.Input(shape=(2,))
    x = tf.keras.layers.Dense(4)(inp)
    return tf.keras.Model(inp, x)


for save_format in ['tf', 'h5']:
    
    print("\nTrying with save_format='{}':\n".format(save_format))
    
    model_with_function = make_model()
    model_with_function.compile(optimizer='adam', loss='mse', metrics=[custom_accuracy])
    model_with_function.save('/tmp/model_with_function' + '.' + save_format, 
                             save_format=save_format)
    
    try:
        new_model = tf.keras.models.load_model('/tmp/model_with_function' + '.' + save_format, 
                                               custom_objects={'custom_accuracy': custom_accuracy}, 
                                               compile=True)
        print("model_with_function loaded with the following metrics:")
        print(new_model.metrics)
    except Exception as e:
        print("model_with_function not loaded with the following error:")
        print(type(e))
        print(e)
        
    model_with_subclass = make_model()
    model_with_subclass.compile(optimizer='adam', loss='mse', metrics=[CustomAccuracy()])
    model_with_subclass.save('/tmp/model_with_subclass' + '.' + save_format, 
                             save_format=save_format)
    
    try:
        new_model = tf.keras.models.load_model('/tmp/model_with_subclass' + '.' + save_format, 
                                               custom_objects={'CustomAccuracy': CustomAccuracy}, 
                                               compile=True)
        print("model_with_function loaded with the following metrics:")
        print(new_model.metrics)
    except Exception as e:
        print("model_with_function not loaded with the following error:")
        print(type(e))
        print(e)
```

I think that my code was already minimal as it just:

1.  created the simplest custom accuracy possible
2. created the simplest MLP possible
3. compiled the MLP with the custom accuracy
4. saved the MLP
5. loaded the MLP

I don't know how I can make it simpler. Except if you want the same piece of code but without the `print` calls and without the `try` and `except` blocks. In this case here it is:
```
import tensorflow as tf

from tensorflow.python.keras.metrics import MeanMetricWrapper
from tensorflow.python.keras.metrics import accuracy

def custom_accuracy(y_true, y_pred):
    return accuracy(y_true, y_pred)

class CustomAccuracy(MeanMetricWrapper):

    def __init__(self, **kwargs):
        super(CustomAccuracy, self).__init__(custom_accuracy, **kwargs)
        
def make_model():
    inp = tf.keras.Input(shape=(2,))
    x = tf.keras.layers.Dense(4)(inp)
    return tf.keras.Model(inp, x)

SAVE_FORMAT = 'h5'
#SAVE_FORMAT = 'tf'

model_with_function = make_model()
model_with_function.compile(optimizer='adam', loss='mse', metrics=[custom_accuracy])
model_with_function.save('/tmp/model_with_function' + '.' + SAVE_FORMAT, 
                         save_format=SAVE_FORMAT)


new_model = tf.keras.models.load_model('/tmp/model_with_function' + '.' + SAVE_FORMAT, 
                                           custom_objects={'custom_accuracy': custom_accuracy}, 
                                           compile=True)

    
#model_with_subclass = make_model()
#model_with_subclass.compile(loss='mse', metrics=[CustomAccuracy()])
#model_with_subclass.save('/tmp/model_with_subclass' + '.' + SAVE_FORMAT, 
#                         save_format=SAVE_FORMAT)
#
#new_model = tf.keras.models.load_model('/tmp/model_with_subclass' + '.' + SAVE_FORMAT, 
#                                           custom_objects={'CustomAccuracy': CustomAccuracy}, 
#                                           compile=True)
```
but you have to manually comment or uncomment some parts if you want to observe all four cases.

Ironically, adding an optimizer for tf2.0.0-beta1 makes the code less minimal.

**Comment by ravikyram** [comment]
@durandg12 

Thanks for the detailed explanation. After adding `optimizer='adam'` in compile call i am able to reproduce the same error message in both  TF 2.0.0 and 2.0.0-beta1. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/17024d04ca0bc9a7e28cec4244715c7a/untitled355.ipynb).Thanks!

**Comment by jvishnuvardhan** [comment]
@durandg12 Thanks for the detailed report. As of now there is no solution available. There are some workarounds suggested [here](https://github.com/tensorflow/tensorflow/issues/32612). There is a PR https://github.com/tensorflow/tensorflow/pull/33229 to resolve an issue similar to this issue. Please follow the PR and test it once it is approved and released in `tf-nightly`. Thanks!

**Comment by durandg12** [comment]
I have reviewed the issue you linked. It seems to be the same problem indeed. I had also found the workaround of loading without compile but as @somedadaism said [this post](https://github.com/tensorflow/tensorflow/issues/32612#issuecomment-534248672) it is not satisfying.

So right now the best workaround is to use a custom function and pass it to the `compile`method and not subclassing `MeanMetricWrapper`.

@jvishnuvardhan I did not try the PR yet, I am not sure how to do it. Am I supposed to create a new virualenv and install `tf-nightly` in it? How do I know when the PR is approved and released?

**Comment by jvishnuvardhan** [comment]
@durandg12 If you have a solution to an issue in Tensorflow, you can raise PR by going [here](https://github.com/tensorflow/tensorflow/pulls). You can edit related TF source code with your solution, test it locally, then checkit into PR. Tensorflow Team will review it and responds. If everything is looking good, then it will be approved and then merged into TF source code. [This](https://www.tensorflow.org/community/contribute) is a very good resource to start contributing. Hope this helps. Thanks!

**Comment by durandg12** [comment]
@jvishnuvardhan my question was more focused on your last sentence, as I know what is a PR in general. My question is how do I do this:

> Please follow the PR and test it once it is approved and released in `tf-nightly`

I see that the PR is actually awaiting review so it is not approved yet. Once it is approved, what steps do I need to follow? Can you confirm that I just have to set a new virtual env up, run `pip install tf-nightly`, and then try my example code?

**Comment by jvishnuvardhan** [comment]
Once it is approved, you don't need to do anything. After approval, it will be merged into `tf-nightly`.  During the approval process, Reviewer will guide you through the steps if any required. Thanks!

**Comment by jvishnuvardhan** [comment]
@durandg12 As of now https://github.com/tensorflow/tensorflow/pull/33229 was approved but not merged. Once it is merged, you can use `tf-nightly` to test it. Thanks!

**Comment by jvishnuvardhan** [comment]
@durandg12 Can you try `tf-nightly` tomorrow as the related PR merged. Please let us know whether it solved your issue or not. Thanks!

**Comment by jvishnuvardhan** [comment]
@durandg12 Looks like `load_model` is working for both the cases when the model is saved in 'h5` format. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/95b9d359762b766dba337e0cfa0c248b/untitled355.ipynb) is the gist. Thanks!

Both the cases are still failing when the model was saved in `tf` format. Thanks!

**Comment by durandg12** [comment]
I have seen your gist, and after installing tf-nightly I have been able to replicate it on my laptop, thank you.

The onlt small difference I see is that locally I ahve an additional warning:
`WARNING: Logging before flag parsing goes to stderr.`

**Comment by ymodak** [comment]
This is fixed latest tf-nightly version `'2.2.0-dev20200123'`. Thanks!

**Comment by durandg12** [comment]
I have tested and the issue is indeed fixed.

But we shall note that the `tf`mode still raises a warning : 
```
WARNING:tensorflow:From /path/to/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1809: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
```

**Comment by grantbrewster** [comment]
Hello! Was this ever solved for saving/loading custom metrics in SavedModel format opposed to .h5?
